internal/s3select/select.go

```diff
@@ -546,6 +546,7 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
 			return false
 		}
 		outputQueue = outputQueue[:0]
+		bufPool.Put(buf)
```

Comment:
We need to add `defer bufPool.Put(buf)` in the beginning and remove all other places before each return.

```diff
diff --git a/internal/s3select/select.go b/internal/s3select/select.go
index 900cc72c5..be93a79f3 100644
--- a/internal/s3select/select.go
+++ b/internal/s3select/select.go
@@ -522,6 +522,7 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
        sendRecord := func() bool {
                buf := bufPool.Get().(*bytes.Buffer)
                buf.Reset()
+               defer bufPool.Put(buf)
 
                for _, outputRecord := range outputQueue {
                        if outputRecord == nil {
@@ -529,12 +530,10 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
                        }
                        before := buf.Len()
                        if err = s3Select.marshal(buf, outputRecord); err != nil {
-                               bufPool.Put(buf)
                                return false
                        }
                        if buf.Len()-before > maxRecordSize {
                                writer.FinishWithError("OverMaxRecordSize", "The length of a record in the input or result is greater than maxCharsPerRecord of 1 MB.")
-                               bufPool.Put(buf)
                                return false
                        }
                }
@@ -542,7 +541,6 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
                if err = writer.SendRecord(buf); err != nil {
                        // FIXME: log this error.
                        err = nil
-                       bufPool.Put(buf)
                        return false
                }
                outputQueue = outputQueue[:0]
```

This is the correct fix.